alpaca/rest.go

```diff
@@ -742,6 +743,155 @@ func (c *Client) GetAsset(symbol string) (*Asset, error) {
 	return &asset, nil
 }
 
+const (
+	optionContractsRequestsMaxLimit = 10000
+)
+
+func setQueryLimit(q url.Values, totalLimit, pageLimit, received, maxLimit int) {
+	limit := 0 // use server side default if unset
+	if pageLimit != 0 {
+		limit = pageLimit
+	}
+	if totalLimit != 0 {
+		remaining := totalLimit - received
+		if remaining <= 0 { // this should never happen
+			return
+		}
+		if (limit == 0 || limit > remaining) && remaining <= maxLimit {
+			limit = remaining
+		}
+	}
+
+	if limit != 0 {
+		q.Set("limit", strconv.Itoa(limit))
+	}
+}
+
+type GetOptionContractsRequest struct {
+	UnderlyingSymbols     string
+	ShowDeliverable       bool
+	Status                OptionStatus
+	ExpirationDate        civil.Date
+	ExpirationDateGTE     civil.Date
+	ExpirationDateLTE     civil.Date
+	RootSymbol            string
+	Type                  OptionType
+	Style                 OptionStyle
+	StrikePriceGTE        decimal.Decimal
+	StrikePriceLTE        decimal.Decimal
+	PennyProgramIndicator bool
+	PageLimit             int
+	TotalLimit            int
+}
+
+// GetOptionContracts returns the list of Option Contracts.
+func (c *Client) GetOptionContracts(req GetOptionContractsRequest) ([]OptionContract, error) {
+	u, err := url.Parse(fmt.Sprintf("%s/%s/options/contracts", c.opts.BaseURL, apiVersion))
+	if err != nil {
+		return nil, err
+	}
+
+	q := u.Query()
+
+	if req.UnderlyingSymbols != "" {
+		q.Set("underlying_symbols", req.UnderlyingSymbols)
+	}
+
+	q.Set("show_deliverables", strconv.FormatBool(req.ShowDeliverable))
+
+	if req.Status == OptionStatusActive || req.Status == OptionStatusInactive {
+		q.Set("status", string(req.Status))
+	}
+
+	defaultDate := civil.Date{}
+	if req.ExpirationDate != defaultDate {
+		q.Set("expiration_date", req.ExpirationDate.String())
+	}
+
+	if req.ExpirationDateGTE != defaultDate {
+		q.Set("expiration_date_gte", req.ExpirationDateGTE.String())
+	}
+
+	if req.ExpirationDateLTE != defaultDate {
+		q.Set("expiration_date_lte", req.ExpirationDateLTE.String())
+	}
+
+	if req.RootSymbol != "" {
+		q.Set("root_symbol", req.RootSymbol)
+	}
+
+	if req.Type == OptionTypeCall || req.Type == OptionTypePut {
+		q.Set("type", string(req.Type))
+	}
+
+	if req.Style == OptionStyleAmerican || req.Style == OptionStyleEuropean {
+		q.Set("style", string(req.Style))
+	}
+
+	defaultStrikePrice := decimal.Decimal{}
+	if req.StrikePriceLTE.Cmp(defaultStrikePrice) != 0 {
+		q.Set("strike_price_lte", req.StrikePriceLTE.String())
+	}
+
+	if req.StrikePriceGTE.Cmp(defaultStrikePrice) != 0 {
+		q.Set("strike_price_gte", req.StrikePriceGTE.String())
+	}
+
+	q.Set("ppind", strconv.FormatBool(req.PennyProgramIndicator))
```

Comment:
This seems to be broken on the server side:
``` shell
$ curl -H "APCA-API-KEY-ID: ${APCA_API_KEY_ID}" -H "APCA-API-SECRET-KEY: ${APCA_API_SECRET_KEY}" \
    "https://paper-api.alpaca.markets/v2/options/contracts?ppind=false"
{"code":40010001,"message":"request parameters are invalid"}
```
I reported this issue internally, but for now, let's work around it by only setting this when it's true, like:
``` go
if req.PennyProgramIndicator {
	q.Set("ppind", "true")
}
```